src/main/java/org/wildfly/security/http/HttpAuthenticator.java

```diff
@@ -272,7 +272,7 @@ private boolean authenticate() throws HttpAuthenticationException {
                             throw log.httpAuthenticationNoSuccessfulResponder();
                         }
                         if (statusSet == false) {
-                            httpExchangeSpi.setStatusCode(OK);
+                            httpExchangeSpi.setStatusCode(FORBIDDEN);
```

Comment:
This seems a big change in behaviour, I believe the previous intent was to default to 'OK' if no mechanism wanted a specific value - shouldn't the FORBIDDEN scenario have been picked up previously?

Reply:
The default OK is (from my observation) used only by FORM mechanism.
For any other mechanism it does not make sense to return OK when authentication was not sufficiently done.
This also correspond to actively used FORBIDDEN on line 281, where is the same situation, just the responder is not used.
Why should using the responder change default status code? It is unexpected from HTTP mechanism implementator point of view.

Reply:
On line 281 the reason we can go straight to FORBIDDEN is because we know authentication was required but not successful or it was attempted but failed in an optional scenario when configured to not ignore failures - added to that we have no responder available to generate a challenge so the only option is to flag the request as forbidden.

Where a responder is used if that responder wants a different status code it is the responders responsibility to set it.

Also on line 263 if a mechanism sets the status code to 200 we deliberately ignore it so we can fall back to 200 on line 275 where no mechanism specific response code has been established.

Reply:
Where a responder is used, it can set any status code it want. I believe empty responder should be equivalent to no responder.
Ad 263: ok, that is problem... back to blueprints...

Reply:
To achieve the fix you want here we may need to look at some form of giving the status codes some form of priority order.

Default to 200 if no status set (Default Behaviour)
If only status requested by a mech is 403 use it.
First mech to request anything other than 200 or 403 wins.

That will retain compatibility with existing mechanisms, especially ones that assume we default to 200.  The only mechanism then that needs a code change is the one you are trying to write a fix for anyway.